### PR TITLE
Add prod op

### PR DESCRIPTION
**Pull request by jacobhinkle**
This adds the `prod` op which performs a product of elements in a tensor along a given set of dimensions.

### CODE REVIEW

File: third_party/nvfuser/csrc/ops/arith.cpp [hunk 1]
```diff
@@ -1347,6 +1347,41 @@ TensorView* sum(
...
+      dtype = DataType::Int;
+    }
+  }
+
+  // Cast input tensor to dtype before the operation is performed
+  if (dtype != DataType::Null) {
+    v1 = optionalCastStrict(dtype, v1)->as<TensorView>();
+  }
+
+  Val* init = nullptr;
```
Comment (rdspring1) [inline third_party/nvfuser/csrc/ops/arith.cpp]: Should this `init` block be refactored like `ops::getMinimumValue` or `ops::getMaximumValue`? I see that `sum` has a similar pattern too.

**Review by rdspring1 — APPROVED** [review]
LGTM

File: third_party/nvfuser/csrc/ops/arith.cpp (hunk 1 above)
Comment (jacobhinkle) [reply third_party/nvfuser/csrc/ops/arith.cpp]: Yeah possibly. We do have `Fusion::oneVal()` and `Fusion::zeroVal()` but they are scalar `Double`s. A simple solution would be to add an optional `dtype` argument to `oneVal` and `zeroVal`. I'll try it.

File: third_party/nvfuser/csrc/ops/arith.cpp [hunk 2]
```diff
@@ -1329,22 +1329,29 @@ TensorView* sum(
...
+    }
   }
 
-  return reductionOp(BinaryOpType::Add, axes, init, v1, keep_dim, dtype);
+  // Cast input tensor to dtype before the operation is performed
+  if (dtype != DataType::Null) {
+    v1 = optionalCastStrict(dtype, v1)->as<TensorView>();
+  }
+
+  auto init = FusionGuard::getCurFusion()->oneVal(v1->getDataType().value());
```
Comment (jacobhinkle) [inline third_party/nvfuser/csrc/ops/arith.cpp:1353]: @rdspring1 we can now just pass a dtype to get a `oneVal` or `zeroVal`. If it is an int or boolean dtype then the cached value is used. Floats and complex dtypes will generate a new constant Val every time.

Comment (rdspring1) [reply third_party/nvfuser/csrc/ops/arith.cpp:1353]: That looks fine.

**Pull request merged**